src/DeviceInput/deviceInputSystem.ts

```diff
@@ -0,0 +1,225 @@
+
+            if (device![inputIndex] != undefined) {
+                return device![inputIndex];
+            }
+            else {
+                throw `Unable to find input ${inputIndex} on device ${deviceName}`;
+            }
+        }
+        else {
+            throw `Unable to find device ${deviceName}`;
+        }
+    }
+
+    /**
+     * When a device is connected, perform user specified function
+     * @param callback Callback function to use when a device is connected
+     */
+    public onDeviceConnected(callback: (deviceName: string) => void): void {
+        this._onDeviceConnected = callback;
+    }
+
+    /**
+     * When a device is disconnected, perform user specified function
+     * @param callback Callback function to use when a device is disconnected
+     */
+    public onDeviceDisconnected(callback: (deviceName: string) => void): void {
+        this._onDeviceDisconnected = callback;
+    }
+
+    // Private functions
+    /**
+     * Add device and inputs to device map
+     * @param deviceName Assigned name of device (may be SN)
+     * @param numberOfInputs Number of input entries to create for given device
+     */
+    private registerDevice(deviceName: string, numberOfInputs: number) {
+        if (!this._inputs.has(deviceName)) {
+            var device: Array<number> = [];
+            for (var i = 0; i < numberOfInputs; i++) {
+                device.push(0);
+            }
+
+            this._inputs.set(deviceName, device);
+        }
+    }
+
+    /**
+     * Given a specific device name, remove that device from the device map
+     * @param deviceName Name of device to be removed
+     */
+    private deregisterDevice(deviceName: string) {
+        this._inputs.delete(deviceName);
+    }
+
+    /**
+     * Handle all actions that come from keyboard interaction
+     */
+    private handleKeyActions() {
+        window.addEventListener("keydown", (evt) => {
+            if (!this._keyboardActive) {
+                this._keyboardActive = true;
+                this.registerDevice(DeviceInputSystem.KEYBOARD_DEVICE, this._maxKeyCodes);
+                this._onDeviceConnected(DeviceInputSystem.KEYBOARD_DEVICE);
+            }
+
+            let kbKey = this._inputs.get(DeviceInputSystem.KEYBOARD_DEVICE);
+            if (kbKey) {
+                kbKey[evt.keyCode] = 1;
+            }
+        });
+
+        window.addEventListener("keyup", (evt) => {
+            let kbKey = this._inputs.get(DeviceInputSystem.KEYBOARD_DEVICE);
+            if (kbKey) {
+                kbKey[evt.keyCode] = 0;
+            }
+        });
+    }
+
+    /**
+     * Handle all actions that come from pointer interaction
+     */
+    private handlePointerActions() {
+        this._elementToAttachTo?.addEventListener("pointermove", (evt) => {
+            if (!this._pointerActive) {
+                this._pointerActive = true;
+                this.registerDevice(DeviceInputSystem.POINTER_DEVICE, this._maxPointerButtons);

```

Comment:
I struggle to understand then how this works with multi touch? each pointer has an id to differenciate them and I was expecting the system to handle it

Reply:
Ah...  Now I'm following the pointer id thing.  As far as the current iteration of the code, it wouldn't support multi-touch but at this step that'd be easy enough to add.